cmd/containerd-shim-runc-v2/manager/manager_linux.go

```diff
@@ -146,44 +200,35 @@ func (manager) Start(ctx context.Context, id string, opts shim.StartOpts) (_ shi
 			break
 		}
 	}
-	address, err := shim.SocketAddress(ctx, opts.Address, grouping)
-	if err != nil {
-		return params, err
-	}
 
-	socket, err := shim.NewSocket(address)
-	if err != nil {
-		// the only time where this would happen is if there is a bug and the socket
-		// was not cleaned up in the cleanup method of the shim or we are using the
-		// grouping functionality where the new process should be run with the same
-		// shim as an existing container
-		if !shim.SocketEaddrinuse(err) {
-			return params, fmt.Errorf("create new shim socket: %w", err)
-		}
-		if shim.CanConnect(address) {
-			params.Address = address
-			return params, nil
-		}
-		if err := shim.RemoveSocket(address); err != nil {
-			return params, fmt.Errorf("remove pre-existing socket: %w", err)
-		}
-		if socket, err = shim.NewSocket(address); err != nil {
-			return params, fmt.Errorf("try create new shim socket 2x: %w", err)
-		}
-	}
+	var sockets []*shimSocket
 	defer func() {
 		if retErr != nil {
-			socket.Close()
-			_ = shim.RemoveSocket(address)
+			for _, s := range sockets {
+				s.Close()
+			}
 		}
 	}()
 
-	f, err := socket.File()
+	s, err := newShimSocket(ctx, opts.Address, grouping, false)
 	if err != nil {
+		if errdefs.IsAlreadyExists(err) {
+			params.Address = s.addr
+			return params, nil
+		}
 		return params, err
 	}
+	sockets = append(sockets, s)
+	cmd.ExtraFiles = append(cmd.ExtraFiles, s.f)
 
-	cmd.ExtraFiles = append(cmd.ExtraFiles, f)
+	if opts.Debug {
+		s, err = newShimSocket(ctx, opts.Address, grouping, true)
```

Comment:
This will create a new debug socket for pprof, which results in shim creating two sockets. Is it prossible to use just one socket instead?
```
[root@devnode1 containerd]# ls /run/containerd/s/
[root@devnode1 containerd]# ctr run -d docker.io/library/nginx:latest mydemo
[root@devnode1 containerd]# ls /run/containerd/s/
0b6efc3510d8ebcb7d27d04dfe4444511fcb6a1fb95411e15b50888cdec86dc3
12bede06714e38de587a8a27a83b967a4a5263ff48c6186a4ec274927a8bc692
[root@devnode1 containerd]# ctr shim --id mydemo pprof goroutines > goroutines.prof
[root@devnode1 containerd]#
```

Reply:
We have to create a separate socket for pprof because
1. The regular endpoint and pprof endpoint do not share protocols (ttrpc vs http)
2. A number of pprof command could potentially block the socket for a long period of time. e.g. the `trace` command can take a `second` parameter and trace the program for an arbitrary timeframe.

Reply:
Thank you for your explanation, I understand now.